opnode/node/server.go

```diff
@@ -0,0 +1,257 @@
+	}
+
+	w.Header().Set("content-type", "application/json")
+	w.WriteHeader(statusCode)
+	enc := json.NewEncoder(w)
+	if err := enc.Encode(res); err != nil {
+		log.Error("error writing rpc response", "err", err)
+		return
+	}
+}
+
+type rpcBlock struct {
+	*types.Header
+	Hash         common.Hash        `json:"hash"`
+	Tranasctions types.Transactions `json:"transactions"`
+}
+
+type getProof struct {
+	Address     common.Address `json:"address"`
+	StorageHash common.Hash    `json:"storage_hash"`
+}
+
+func (s *rpcServer) outputAtBlock(ctx context.Context, req *RPCReq) *RPCRes {
+	var params []string
+	if err := json.Unmarshal(req.Params, &params); err != nil {
+		return NewRPCErrorRes(req.ID, err)
+	}
+	if len(params) != 1 {
+		return NewRPCErrorRes(req.ID, fmt.Errorf("invalid number of parameters"))
+	}
+
+	blockTag := params[0]
+
+	if blockTag == "safe" {
+		// TODO: handle this via getBlockNumber() - FINALIZED_PERIOD
+	}
+
+	var block *rpcBlock
+	err := s.l2RPCClient.CallContext(ctx, &block, "eth_getBlockByNumber", blockTag, false)
+	if err == nil {
+		return NewRPCErrorRes(req.ID, err)
+	}
+	if block == nil {
+		return NewRPCErrorRes(req.ID, ethereum.NotFound)
+	}
+
+	var getProofResponse *getProof
+	err = s.l2RPCClient.CallContext(ctx, &getProofResponse, "eth_getProof", s.withdrawalContractAddr, nil, blockTag)
+	if err != nil {
+		return NewRPCErrorRes(req.ID, err)
+	} else if getProofResponse == nil {
+		return NewRPCErrorRes(req.ID, ethereum.NotFound)
+	}
+
+	var random l2.Bytes32
+	copy(random[:], block.Difficulty.Bytes())
+
+	txs := make([]l2.Data, 0, len(block.Tranasctions))
+	for _, tx := range block.Tranasctions {
+		w := new(bytes.Buffer)
+		if err := tx.EncodeRLP(w); err != nil {
+			return NewRPCErrorRes(req.ID, fmt.Errorf("invalid transaction found in block"))
+		}
+		txs = append(txs, w.Bytes())
+	}
+
+	payload := &l2.ExecutionPayload{
+		ParentHash:    block.ParentHash,
+		FeeRecipient:  block.Coinbase,
+		StateRoot:     l2.Bytes32(block.Root),
+		ReceiptsRoot:  l2.Bytes32(block.ReceiptHash),
+		LogsBloom:     l2.Bytes256(block.Bloom),
+		Random:        random,
+		BlockNumber:   hexutil.Uint64(block.Number.Uint64()),
+		GasLimit:      hexutil.Uint64(block.GasLimit),
+		GasUsed:       hexutil.Uint64(block.GasUsed),
+		Timestamp:     hexutil.Uint64(block.Time),
+		ExtraData:     block.Extra,
+		BaseFeePerGas: *uint256.NewInt(block.BaseFee.Uint64()),
+		BlockHash:     block.Hash,
+		Transactions:  txs,
+	}
+
+	l2Output := &l2.L2Output{
+		StateRoot:              payload.StateRoot,
+		WithdrawalStorageRoot:  l2.Bytes32(getProofResponse.StorageHash),
+		LatestBlock:            payload,
+		HistoryAccumulatorRoot: l2.Bytes32{}, // unused. zeroed
+		Extension:              l2.Bytes32{}, // unused. zeroed
+	}
+
+	var buf bytes.Buffer
+	bufWriter := bufio.NewWriter(&buf)
+	l2OutputSSZ := zssz.GetSSZ((*l2.L2Output)(nil))
+	if _, err := zssz.Encode(bufWriter, &l2Output, l2OutputSSZ); err != nil {
+		return NewRPCErrorRes(req.ID, err)
+	}
```

Comment:
please don't use ZSSZ. Although it works and it's fast, and was fuzzed at some point, I have not updated it for 2 years. Instead, use https://github.com/protolambda/ztyp if you need recent SSZ code (spoiler: we don't need it at all, for now). It's a bit slower, but doesn't use unsafe Go, and in https://github.com/protolambda/zrnt/ it's widely tested against the ethereum standard tests.

Also, encoding and then hashing the SSZ data effectively **does nothing useful for the verifier**. The original intention was to use **hash-tree-root**, a **different** feature of SSZ than encoding, that merkleizes the structure, enabling an efficient proof to the contents of the block.

Not sure if I want the dependency at all though. We are trying to simplify the withdrawals format, and don't need SSZ for that.

If we are going with the simplified output-root format let's just do something like:
```
var buf bytes.Buffer
buf.Write([]byte{versionPrefixByte})  // or we can do a 32 bytes version, if that is nicer to proof in EVM
buf.Write(l2Output.StateRoot[:])
buf.Write(l2Output.WithdrawalStorageRoot[:])
buf.Write(l2Output.BlockHash[:])  // not the ssz hash-tree-root, just regular eth1 block-hash
hash := crypto.Keccak256(buf.Bytes())
```

cc @maurelian we've to update the spec with the simplified version, so we don't end up implementing the wrong thing (sorry @Inphi !)